starters/nextjs-starter-ts/hooks/usePagination.ts

```diff
@@ -0,0 +1,55 @@
+import {
+  ArticleWithoutContent,
+  PaginatedArticle,
+} from "@pantheon-systems/pcc-sdk-core/types";
+import { useEffect, useState } from "react";
+
+interface Props {
+  cursor?: string;
+  pageSize: number;
+  initialArticles?: PaginatedArticle[] | ArticleWithoutContent[];
+}
+
+export function usePagination({ cursor, initialArticles, pageSize }: Props) {
+  const [currentCursor, setCurrentCursor] = useState(cursor);
+  const [articlePages, setArticlePages] = useState(
+    initialArticles ? [initialArticles] : [],
+  );
+  const [fetching, setFetching] = useState(false);
+  const [currentPage, setCurrentPage] = useState(0);
+
+  useEffect(() => {
+    (async () => {
+      if (articlePages[currentPage]) return;
+
+      setFetching(true);
+      const response = await fetch(
+        "/api/utils/paginate/?" +
+          new URLSearchParams({
+            pageSize: pageSize.toString(),
+            cursor: currentCursor || "",
+          }).toString(),
+      );
+      const { data, newCursor } = (await response.json()) as {
+        data: PaginatedArticle[];
+        newCursor: string;
+      };
+      setFetching(false);
+      setArticlePages((prev) => {
+        const result = [...prev];
+        result[currentPage] = data;
+        return result;
+      });
+      setCurrentCursor(newCursor);
+    })();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [currentPage]);
```

Comment:
Use react-query. Note that they changed their name to @tanstack/react-query (https://tanstack.com/query/latest).

I recommend to start on the pages-router projects first to get a handle on it before implementing on approuter, since there may be some extra work on the approuter side to avoid multiple query contexts issue.

Reply:
Integrated `useSWR` hook instead of react-query since we are already using `swr` in the starter kit.
I have not added it for approuter starter since it's using server action instead of the `fetch` call to paginate route.